src/try_fold.rs

```diff
@@ -34,9 +34,12 @@ impl<'a, I, O: Clone + 'a, E> TryFold<'a, I, O, E> {
   /// Returns a combined `And` structure that represents the sequential folding operation.
   pub fn and(self, other: TryFold<'a, I, O, E>) -> Self {
     TryFold(Box::new(move |input, state| {
-      self
-        .try_fold(input, state.clone())
-        .fold(|state| other.try_fold(input, state), other.try_fold(input, state))
+      let r = self.try_fold(input, state.clone()).to_result();
+
+      match r {
+        Ok(state) => other.try_fold(input, state),
+        Err(error) => Valid::<O, E>::from_validation_err(error).and(other.try_fold(input, state)),
+      }
```

Comment:
As a general rule of thumb, we should always use the most fundamental operator that's available on the domain. In this case, Valid has the `fold` operator and we should leverage it, instead of using the result. This has two advantages - 
1. Your implementation remains unaffected when the `Result` type is dropped from Valid.
2. The code is more concise, and maintains correctness without duplicating lot of logic.
3. Optimizing `fold` will optimize every operator that's built on it.